File: android/app/src/main/java/com/videocallwidget/TodoAppWidgetProvider.java
package com.videocallwidget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import com.google.gson.Gson;


public class TodoAppWidgetProvider extends AppWidgetProvider {

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        if (intent.getAction().equals(AppWidgetManager.ACTION_APPWIDGET_UPDATE)) {
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            ComponentName thisAppWidget = new ComponentName(context.getPackageName(), TodoAppWidgetProvider.class.getName());
            int[] appWidgetIds = appWidgetManager.getAppWidgetIds(thisAppWidget);
            onUpdate(context, appWidgetManager, appWidgetIds);
        }
    }
    
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        String todosJson = preferences.getString("todos", "[]");

        // Convert JSON string to list of todos
        Gson gson = new Gson();
        String[] todosArray = gson.fromJson(todosJson, String[].class);

        for (int appWidgetId : appWidgetIds) {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.todo_app_widget);

            Intent intent = new Intent(context, MainActivity.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
            views.setOnClickPendingIntent(R.id.todo_text, pendingIntent);

            // Update widget view with todos
            views.setTextViewText(R.id.todo_text, String.join("\n", todosArray));

            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }
}

File: android/app/src/main/java/com/videocallwidget/TodoAppWidgetProvider.kt
package com.videocallwidget

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import com.google.gson.Gson

class TodoAppWidgetProvider : AppWidgetProvider() {

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        val preferences = context.getSharedPreferences("com.videocallwidget.preferences", Context.MODE_PRIVATE)
        val todosJson = preferences.getString("todos", "[]")

        // Convert JSON string to list of todos
        val gson = Gson()
        val todosArray: Array<String> = gson.fromJson(todosJson, Array<String>::class.java)

        for (appWidgetId in appWidgetIds) {
            val views = RemoteViews(context.packageName, R.layout.todo_app_widget)
            // Update widget view with todos
            views.setTextViewText(R.id.todo_text, todosArray.joinToString("\n"))

            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        if ("com.videocallwidget.UPDATE_WIDGET" == intent.action) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val thisAppWidget = ComponentName(context.packageName, TodoAppWidgetProvider::class.java.name)
            val appWidgetIds = appWidgetManager.getAppWidgetIds(thisAppWidget)
            onUpdate(context, appWidgetManager, appWidgetIds)
        }
    }
}
